tsl/src/remote/async.c

```diff
@@ -512,6 +512,137 @@ async_request_wait_ok_result(AsyncRequest *req)
 	return res;
 }
 
+/*
+ * Get the result of an async request during cleanup.
+ *
+ * Cleanup is typically necessary for a query that is being interrupted by
+ * transaction abort, or a query that was initiated as part of transaction
+ * abort to get the remote side back to the appropriate state.
+ *
+ * endtime is the time at which we should give up and assume the remote
+ * side is dead.
+ *
+ * An AsyncReponse is always returned, indicating last PGresult received,
+ * a timeout, or error.
+ */
+AsyncResponse *
+async_request_cleanup_result(AsyncRequest *req, TimestampTz endtime)
+{
+	volatile bool timed_out = false;
+	volatile bool conn_issue = false;
+	PGresult *volatile last_res = NULL;
+	const TSConnection *conn = async_request_get_connection(req);
+	PGconn *pg_conn = remote_connection_get_pg_conn(conn);
+
+	switch (req->state)
+	{
+		case DEFERRED:
+			if (remote_connection_is_processing(req->conn))
+				return async_response_error_create("request already in progress");
+
+			req = async_request_send_internal(req, WARNING);
+
+			if (req == NULL)
+				return async_response_error_create("failed to send deferred request");
+
+			Assert(req->state == EXECUTING);
+			break;
+		case EXECUTING:
+			break;
+		case COMPLETED:
+			return async_response_error_create("request already completed");
+	}
+
+	/* In what follows, do not leak any PGresults on an error. */
+	PG_TRY();
+	{
+		for (;;)
+		{
+			PGresult *res;
+
+			while (PQisBusy(pg_conn))
+			{
+				int wc;
+				TimestampTz now = GetCurrentTimestamp();
+				long secs;
+				int microsecs;
+				long cur_timeout;
```

Comment:
It is usually quite useful to indicate the unit of a timeout for variables that contain timeouts. Avoids confusion when the variable is used.
```suggestion
				long cur_timeout_msec;
```